src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -92,6 +92,25 @@ public static string GetDirectoryName(string path)
             return null;
         }
 
+        public static ReadOnlySpan<char> GetDirectoryName(ReadOnlySpan<char> path)
+        {
+            if (path.IsEmpty)
+                return path;
+
+            path = PathInternal.NormalizeDirectorySeparators(path);
+            int root = PathInternal.GetRootLength(new string(path));
```

Comment:
By allocating a string, you're partly defeating the purpose of this change, which is to offer a way to do Path operations without allocating strings.

Instead it should call the GetRootLength that takes a char*. Use DangerousGetPinnableReference() to get that.

Optionally `GetRootLength(char* ..` could be changed to use `ReadOnlySpan<char>` instead. Depends how far you want to go down the rabbit hole.

Reply:
Based on @jkotas note above I suggest indeed changing GetRootLength to use ROS instead of char*.